File: tools/check_dependabot_policy.py
#!/usr/bin/env python3
"""Validate the bounded Dependabot update policy without third-party YAML tooling."""

from pathlib import Path
import re


CONFIG = Path(__file__).parents[1] / ".github" / "dependabot.yml"
EXPECTED = {
    ("cargo", "/"),
    ("cargo", "/py"),
    ("cargo", "/crates/wasm"),
    ("cargo", "/crates/cloud"),
    ("cargo", "/containers/verify-runner"),
    ("npm", "/sdk/ts"),
    ("npm", "/sdk/mcp"),
    ("npm", "/web"),
    ("npm", "/docs/site"),
    ("npm", "/services/cloud"),
    ("npm", "/tools/license-issuer"),
    ("github-actions", "/"),
}


def fail(message: str) -> None:
    raise SystemExit(f"dependabot policy: {message}")


def main() -> None:
    text = CONFIG.read_text(encoding="utf-8")
    if not text.startswith("version: 2\nupdates:\n"):
        fail("expected a version 2 updates document")

    blocks = text.split("\n  - package-ecosystem: ")[1:]
    found: list[tuple[str, str]] = []
    for block in blocks:
        ecosystem, separator, rest = block.partition("\n")
        if not separator:
            fail("truncated package-ecosystem entry")
        directory_match = re.search(r"^    directory: (\S+)\s*$", rest, re.MULTILINE)
        if directory_match is None:
            fail(f"{ecosystem}: missing directory")
        key = (ecosystem.strip(), directory_match.group(1))
        found.append(key)

        limits = re.findall(r"^    open-pull-requests-limit: (\d+)\s*$", rest, re.MULTILINE)
        if limits != ["2"]:
            fail(f"{key}: expected exactly one open-pull-requests-limit: 2")
        if not re.search(
            r"^    groups:\n      non-major:\n        patterns: \[\"\*\"\]\n"
            r"        update-types: \[\"minor\", \"patch\"\]\s*$",
            rest,
            re.MULTILINE,
        ):
            fail(f"{key}: expected the canonical minor/patch group")

    if len(found) != len(set(found)):
        fail("duplicate ecosystem/directory entry")
    actual = set(found)
    if actual != EXPECTED:
        fail(f"entry mismatch; missing={sorted(EXPECTED - actual)}, extra={sorted(actual - EXPECTED)}")

    print(f"dependabot policy: ok ({len(found)} bounded update roots)")


if __name__ == "__main__":
    main()

File: tools/check_dependabot_policy.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "yaml"

config_path = File.expand_path("../.github/dependabot.yml", __dir__)
expected = [
  ["cargo", "/"],
  ["cargo", "/py"],
  ["cargo", "/crates/wasm"],
  ["cargo", "/crates/cloud"],
  ["cargo", "/containers/verify-runner"],
  ["npm", "/sdk/ts"],
  ["npm", "/sdk/mcp"],
  ["npm", "/web"],
  ["npm", "/docs/site"],
  ["npm", "/services/cloud"],
  ["npm", "/tools/license-issuer"],
  ["github-actions", "/"]
].freeze

def fail_policy(message)
  warn "dependabot policy: #{message}"
  exit 1
end

config = YAML.safe_load(
  File.read(config_path, encoding: "UTF-8"),
  permitted_classes: [],
  permitted_symbols: [],
  aliases: false
)
fail_policy("expected a YAML mapping") unless config.is_a?(Hash)
fail_policy("expected version: 2") unless config["version"] == 2

updates = config["updates"]
fail_policy("updates must be a list") unless updates.is_a?(Array)

found = updates.map.with_index do |entry, index|
  fail_policy("updates[#{index}] must be a mapping") unless entry.is_a?(Hash)

  ecosystem = entry["package-ecosystem"]
  directory = entry["directory"]
  key = [ecosystem, directory]
  fail_policy("#{key}: open-pull-requests-limit must equal 2") unless entry["open-pull-requests-limit"] == 2

  groups = entry["groups"]
  fail_policy("#{key}: groups must contain only non-major") unless groups.is_a?(Hash) && groups.keys == ["non-major"]
  non_major = groups["non-major"]
  unless non_major.is_a?(Hash) && non_major["patterns"] == ["*"] && non_major["update-types"] == ["minor", "patch"]
    fail_policy("#{key}: expected the canonical minor/patch group")
  end

  key
end

fail_policy("duplicate ecosystem/directory entry") unless found.length == found.uniq.length
missing = expected - found
extra = found - expected
fail_policy("entry mismatch; missing=#{missing.inspect}, extra=#{extra.inspect}") unless missing.empty? && extra.empty?

puts "dependabot policy: ok (#{found.length} bounded version-update roots)"
